src/ToolBox/SOS/NETCore/SymbolReader.cs

```diff
@@ -0,0 +1,679 @@
+            {
+                peStream = new TargetStream(loadedPeAddress, loadedPeSize, readMemory);
+            }
+            TargetStream pdbStream = null;
+            if (inMemoryPdbAddress != IntPtr.Zero)
+            {
+                pdbStream = new TargetStream(inMemoryPdbAddress, inMemoryPdbSize, readMemory);
+            }
+            OpenedReader openedReader = GetReader(assemblyPath, isFileLayout, peStream, pdbStream);
+            if (openedReader == null)
+                return IntPtr.Zero;
+
+            GCHandle gch = GCHandle.Alloc(openedReader);
+            return GCHandle.ToIntPtr(gch);
+        }
+
+        /// <summary>
+        /// Cleanup and dispose of symbol reader handle
+        /// </summary>
+        /// <param name="symbolReaderHandle">symbol reader handle returned by LoadSymbolsForModule</param>
+        internal static void Dispose(IntPtr symbolReaderHandle)
+        {
+            Debug.Assert(symbolReaderHandle != IntPtr.Zero);
+            try
+            {
+                GCHandle gch = GCHandle.FromIntPtr(symbolReaderHandle);
+                ((OpenedReader)gch.Target).Dispose();
+                gch.Free();
+            }
+            catch
+            {
+            }
+        }
+
+        /// <summary>
+        /// Returns method token and IL offset for given source line number.
+        /// </summary>
+        /// <param name="symbolReaderHandle">symbol reader handle returned by LoadSymbolsForModule</param>
+        /// <param name="fileName">source file name</param>
+        /// <param name="lineNumber">source line number</param>
+        /// <param name="methodToken">method token return</param>
+        /// <param name="ilOffset">IL offset return</param>
+        /// <returns> true if information is available</returns>
+        internal static bool ResolveSequencePoint(IntPtr symbolReaderHandle, string fileName, int lineNumber, out int methodToken, out int ilOffset)
+        {
+            Debug.Assert(symbolReaderHandle != IntPtr.Zero);
+            methodToken = 0;
+            ilOffset = 0;
+
+            GCHandle gch = GCHandle.FromIntPtr(symbolReaderHandle);
+            MetadataReader reader = ((OpenedReader)gch.Target).Reader;
+
+            try
+            {
+                foreach (MethodDebugInformationHandle methodDebugInformationHandle in reader.MethodDebugInformation)
+                {
+                    MethodDebugInformation methodDebugInfo = reader.GetMethodDebugInformation(methodDebugInformationHandle);
+                    SequencePointCollection sequencePoints = methodDebugInfo.GetSequencePoints();
+                    foreach (SequencePoint point in sequencePoints)
+                    {
+                        string sourceName = reader.GetString(reader.GetDocument(point.Document).Name);
+                        if (Path.GetFileName(sourceName) == Path.GetFileName(fileName) && point.StartLine == lineNumber)
```

Comment:
Would it make sense to do the line number check before the filename check since it should be cheaper and the chances of a collision would probably be lower?

BTW, I may be off base here, but if we see that one sequence point in a methodDebugInfo has a filename that doesn't match what we're looking for, do we need to go through the remaining sequence points in that methodDebugInfo?